### PR TITLE
Add I/O priority

### PR DESCRIPTION
**Pull request by utam0k**
https://github.com/opencontainers/runtime-spec/pull/1191

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
Could you squash the commits?

**Comment by utam0k** [comment]
@AkihiroSuda Sure. Done.

**Review by AkihiroSuda — APPROVED** [review]

File: libcontainer/configs/validate/validator.go [hunk 1]
```diff
@@ -286,3 +287,15 @@ func isHostNetNS(path string) (bool, error) {
 
 	return (st1.Dev == st2.Dev) && (st1.Ino == st2.Ino), nil
 }
+
+func ioPriority(config *configs.Config) error {
+	ioPriroty := config.IOPriority
+	if ioPriroty == nil {
```
Comment (kolyshkin) [inline libcontainer/configs/validate/validator.go:293]: nit: your variable name has a typo.

nit: you can just do `if config.IOPriority == nil` instead of adding a variable which you don't use below.

File: libcontainer/configs/validate/validator_test.go [hunk 2]
```diff
@@ -387,3 +387,32 @@ func TestValidateMounts(t *testing.T) {
...
+		isErr    bool
+		priority int
+	}{
+		{isErr: false, priority: 0},
+		{isErr: false, priority: 7},
+		{isErr: true, priority: -1},
+	}
+
+	for _, tc := range testCases {
+		ioPriroty := configs.IOPriority{
```
Comment (kolyshkin) [inline libcontainer/configs/validate/validator_test.go:402]: same here, typo in a variable name

File: libcontainer/configs/validate/validator_test.go [hunk 3]
```diff
@@ -387,3 +387,32 @@ func TestValidateMounts(t *testing.T) {
...
+			Priority: tc.priority,
+		}
+		config := &configs.Config{
+			Rootfs:     "/var",
+			IOPriority: &ioPriroty,
+		}
+
+		err := Validate(config)
+		if tc.isErr && err == nil {
+			t.Errorf("iopriroty: %d, expected error, got nil", tc.priority)
```
Comment (kolyshkin) [inline libcontainer/configs/validate/validator_test.go:412]: same here, typo in the error message

File: libcontainer/configs/validate/validator_test.go [hunk 4]
```diff
@@ -387,3 +387,32 @@ func TestValidateMounts(t *testing.T) {
...
+		}
+		if !tc.isErr && err != nil {
+			t.Errorf("iopriroty: %d, expected nil, got error %v", tc.priority, err)
```
Comment (kolyshkin) [inline libcontainer/configs/validate/validator_test.go:415]: same here

File: libcontainer/standard_init_linux.go [hunk 5]
```diff
@@ -159,6 +160,13 @@ func (l *linuxStandardInit) Init() error {
 			return &os.SyscallError{Syscall: "prctl(SET_NO_NEW_PRIVS)", Err: err}
 		}
 	}
+
+	if l.config.Config.IOPriority != nil {
+		if err := utils.SetIOPriority(l.config.Config.IOPriority); err != nil {
+			return fmt.Errorf("unable to apply iopriority: %w", err)
```
Comment (kolyshkin) [inline libcontainer/standard_init_linux.go:166]: I don't think the err returned from `utils.SetIOPriority` needs to be wrapped

File: libcontainer/process_linux.go [hunk 6]
```diff
@@ -80,6 +80,12 @@ func (p *setnsProcess) signal(sig os.Signal) error {
 
 func (p *setnsProcess) start() (retErr error) {
 	defer p.messageSockPair.parent.Close()
+	if p.process.IOPriority != nil {
+		if err := utils.SetIOPriority(p.process.IOPriority); err != nil {
+			return fmt.Errorf("unable to apply iopriority: %w", err)
```
Comment (kolyshkin) [inline libcontainer/process_linux.go:85]: I don't think the err returned from `utils.SetIOPriority` needs to be wrapped

File: tests/integration/ioprio.bats [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+	# Create a container with a specific I/O priority
+	update_config '.process.ioPriority = {"class": "IOPRIO_CLASS_BE", "priority": 4}'
+
+	runc run -d --console-socket "$CONSOLE_SOCKET" test_ioprio
+	[ "$status" -eq 0 ]
+
+	# Check the init process
+	runc exec test_ioprio ionice -p 1
+	[ "$status" -eq 0 ]
+	echo "${output}" | grep -q "best-effort: prio 4"
```
Comment (kolyshkin) [inline tests/integration/ioprio.bats:23]: Slightly better to use something like `[[ "$output" = *'best-effort: prio 4'* ]]`

File: tests/integration/ioprio.bats [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+	# Check the process made from the exec command
+	runc exec test_ioprio ionice
+	[ "$status" -eq 0 ]
+	echo "${output}" | grep -q "best-effort: prio 4"
```
Comment (kolyshkin) [inline tests/integration/ioprio.bats:28]: same here

File: tests/integration/ioprio.bats [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+	runc run -d --console-socket "$CONSOLE_SOCKET" test_ioprio
+	[ "$status" -eq 0 ]
+
+	# Check the init process
+	runc exec test_ioprio ionice -p 1
+	[ "$status" -eq 0 ]
+	echo "${output}" | grep -q "best-effort: prio 4"
+
+	# Check the process made from the exec command
```
Comment (kolyshkin) [inline tests/integration/ioprio.bats:25]: nit: missing period at EOL

File: utils_linux.go [hunk 10]
```diff
@@ -61,6 +61,12 @@ func newProcess(p specs.Process) (*libcontainer.Process, error) {
 		lp.ConsoleHeight = uint16(p.ConsoleSize.Height)
 	}
 
+	if p.IOPriority != nil {
+		lp.IOPriority = &configs.IOPriority{}
+		lp.IOPriority.Class = p.IOPriority.Class
+		lp.IOPriority.Priority = p.IOPriority.Priority
+	}
```
Comment (kolyshkin) [inline utils_linux.go:68]: You can initialize in a single statement, i.e.
```go
lp.IOPriority = &configs.IOPriority{
	Class: p.IOPriority.Class
	Priority: p.IOPriority.Priority
}
```

It's a bit less repetitive this way.

**Review by kolyshkin — CHANGES_REQUESTED** [review]
left a few comments, mostly nits

File: go.mod [hunk 11]
```diff
@@ -12,7 +12,7 @@ require (
 	github.com/godbus/dbus/v5 v5.1.0
 	github.com/moby/sys/mountinfo v0.6.2
 	github.com/mrunalp/fileutils v0.5.0
-	github.com/opencontainers/runtime-spec v1.1.0-rc.2
+	github.com/opencontainers/runtime-spec v1.1.0-rc.3
```
[the remark this answers is not in the document]
Comment (AkihiroSuda) [reply go.mod:15]: Please update https://github.com/opencontainers/runc/blob/main/docs/spec-conformance.md too

[the remark this answers is not in the document]
Comment (utam0k) [reply go.mod]: Sure!

File: docs/terminals.md [hunk 12]
```diff
@@ -58,7 +58,7 @@ you use `runc` directly in something like a `systemd` unit file. To disable
 this `LISTEN_FDS`-style passing just unset `LISTEN_FDS`.
 
 **Be very careful when passing file descriptors to a container process.** Due
-to some Linux kernel (mis)features, a container with access to certain types of
+to some Linux kernel misfeatures, a container with access to certain types of
```
Comment (utam0k) [inline docs/terminals.md:61]: We need to fix it to pass the checking code spell.

**Review by AkihiroSuda — APPROVED** [review]

File: libcontainer/configs/validate/validator.go [hunk 13]
```diff
@@ -286,3 +287,14 @@ func isHostNetNS(path string) (bool, error) {
 
 	return (st1.Dev == st2.Dev) && (st1.Ino == st2.Ino), nil
 }
+
+func ioPriority(config *configs.Config) error {
```
Comment (lengrongfu) [inline libcontainer/configs/validate/validator.go:291]: we wheather need validator io priority class?

**Comment by AkihiroSuda** [comment]
Needs rebase

**Comment by kolyshkin** [comment]
@utam0k can you please rebase?

**Comment by utam0k** [comment]
@AkihiroSuda @kolyshkin 👍 Done

File: libcontainer/utils/utils_unix.go [hunk 14]
```diff
@@ -67,3 +70,22 @@ func NewSockPair(name string) (parent *os.File, child *os.File, err error) {
...
+)
+
+func SetIOPriority(ioprio *configs.IOPriority) error {
+	class, ok := configs.IOPrioClassMapping[ioprio.Class]
+	if !ok {
+		return fmt.Errorf("invalid io priority class: %s", ioprio.Class)
+	}
+
+	// Combine class and priority into a single value
+	iop := (class << 13) | ioprio.Priority
```
Comment (lifubang) [inline libcontainer/utils/utils_unix.go:85]: Could you please put the official reference doc here? I saw a different method in https://github.com/torvalds/linux/blob/v5.18/include/uapi/linux/ioprio.h#L5-L17

**Comment by lifubang** [comment]
@utam0k can you please rebase?

File: libcontainer/configs/config.go [hunk 15]
```diff
@@ -216,6 +216,20 @@ type Config struct {
...
+
+var IOPrioClassMapping = map[specs.IOPriorityClass]int{
+	specs.IOPRIO_CLASS_RT:   1,
+	specs.IOPRIO_CLASS_BE:   2,
+	specs.IOPRIO_CLASS_IDLE: 3,
+}
+
+type IOPriority struct {
+	Class    specs.IOPriorityClass `json:"class"`
+	Priority int                   `json:"priority"`
```
[the remark this answers is not in the document]
Comment (lifubang) [reply libcontainer/configs/config.go:232]: But this is not a big problem, I think upstream projects will always provide a value if `ioPriority` provided. I just want to suggest we should consider zero in numeric type field in `runtime-spec` project.

File: libcontainer/utils/utils_unix.go [hunk 16]
```diff
@@ -9,6 +9,9 @@ import (
 	"os"
 	"strconv"
 	"sync"
+	"syscall"
```
Comment (lifubang) [inline libcontainer/utils/utils_unix.go:12]: According to https://github.com/opencontainers/runc/pull/3956#discussion_r1281896310 , can we switch to `x/sys/unix`?

File: libcontainer/utils/utils_unix.go (hunk 14 above)
Comment (lifubang) [reply libcontainer/utils/utils_unix.go:117]: Do you know why linux do a more complex operation in `IOPRIO_PRIO_VALUE`, although the result is the same as you here `(class << 13) | ioprio.Priority`.

**Comment by lifubang** [comment]
Another question, shall we need to add a support to update the IO priority value for a started container in `runc update`?

File: libcontainer/utils/utils_unix.go (hunk 16 above)
Comment (lifubang) [reply libcontainer/utils/utils_unix.go:12]: @utam0k I think if this one is done, this PR could be merged. Do you have a time to deal with it.

Comment (utam0k) [reply libcontainer/utils/utils_unix.go:12]: Sure, I will be able to do it next weekend.

**Comment by utam0k** [comment]
> Another question, shall we need to add a support to update the IO priority value for a started container in `runc update`?

hmm... At least, OCI Spec Runtime doesn't specify about it. WDYT?

File: libcontainer/utils/utils_unix.go (hunk 16 above)
Comment (utam0k) [reply libcontainer/utils/utils_unix.go]: @lifubang Done

File: docs/terminals.md (hunk 12 above)
Comment (AkihiroSuda) [reply docs/terminals.md:61]: Doesn't seem relevant to the PR

Comment (utam0k) [reply docs/terminals.md:61]: But CI won't allow this word... if I edit this file, CI won't go through unless I fix this part.

**Comment by lifubang** [comment]
Please squash your commits to one, thanks.

**Comment by lifubang** [comment]
Sorry, please rebase and squash your commits to one, the CI will be green.

**Comment by utam0k** [comment]
Sure

**Comment by utam0k** [comment]
@lifubang Done

**Review by lifubang — APPROVED** [review]

**Comment by lifubang** [comment]
@kolyshkin @AkihiroSuda PTAL, Merging is blocked. I think this one can be merged now.

**Review by AkihiroSuda — APPROVED** [review]

**Comment by AkihiroSuda** [comment]
Needs rebase

**Comment by utam0k** [comment]
@AkihiroSuda Done ✅

**Comment by kolyshkin** [comment]
> libcontainer/configs/validate/validator_test.go:620:2: expected declaration, found testCases
@utam0k seems there's an error in this file

**Comment by kolyshkin** [comment]
@utam0k also you have to do `pull --rebase` and avoid merge commits.

**Comment by utam0k** [comment]
@kolyshkin Thanks, done

**Comment by AkihiroSuda** [comment]
` Error: File is not `gofumpt`-ed (gofumpt)`
https://github.com/opencontainers/runc/actions/runs/6533097503/job/17737584357?pr=3783

**Comment by lifubang** [comment]
This PR can be merged once you do a rebase and resolve the conflicts, and I can take over this PR if you are in busy status.

File: docs/terminals.md (hunk 12 above)
Comment (kolyshkin) [reply docs/terminals.md:61]: Please remove this hunk. We'll deal with CI later.

File: libcontainer/process_linux.go [hunk 17]
```diff
@@ -120,6 +120,13 @@ func (p *setnsProcess) signal(sig os.Signal) error {
 
 func (p *setnsProcess) start() (retErr error) {
 	defer p.comm.closeParent()
+
+	if p.process.IOPriority != nil {
```
Comment (kolyshkin) [inline libcontainer/process_linux.go:124]: It would be way better to add
```go
if prio == nil {
   return nil
}
```

to the beginning of `utils.SetIOPriority`, and remove the condition here and in other places that call this

File: libcontainer/specconv/spec_linux.go [hunk 18]
```diff
@@ -498,6 +498,13 @@ func CreateLibcontainerConfig(opts *CreateOpts) (*configs.Config, error) {
 			s := *spec.Process.Scheduler
 			config.Scheduler = &s
 		}
+
+		if spec.Process.IOPriority != nil {
+			config.IOPriority = &configs.IOPriority{
+				Class:    spec.Process.IOPriority.Class,
+				Priority: spec.Process.IOPriority.Priority,
+			}
```
Comment (kolyshkin) [inline libcontainer/specconv/spec_linux.go:506]: This can be done in a simpler way (see the code just above). Something like

```go
p := *spec.Process.IOPriority
config.IOPriority = &p
```

File: libcontainer/process_linux.go [hunk 19]
```diff
@@ -120,6 +120,13 @@ func (p *setnsProcess) signal(sig os.Signal) error {
...
+		if err := utils.SetIOPriority(p.process.IOPriority); err != nil {
```
Comment (kolyshkin) [inline libcontainer/process_linux.go:125]: We don't need `SetIOPriority` anywhere else, so maybe don't move it to `utils`.

File: utils_linux.go [hunk 20]
```diff
@@ -66,6 +66,13 @@ func newProcess(p specs.Process) (*libcontainer.Process, error) {
 		lp.Scheduler = &s
 	}
 
+	if p.IOPriority != nil {
+		lp.IOPriority = &configs.IOPriority{
+			Class:    p.IOPriority.Class,
+			Priority: p.IOPriority.Priority,
+		}
```
Comment (kolyshkin) [inline utils_linux.go:73]: same as https://github.com/opencontainers/runc/pull/3783/commits/cff9be20fcd25ea413396d56b7203f1f6ab5b15a#r1439998156

File: libcontainer/specconv/spec_linux.go (hunk 18 above)
Comment (utam0k) [reply libcontainer/specconv/spec_linux.go:506]: Maybe it will be possible with `type` 👍

**Comment by kolyshkin** [comment]
@utam0k please avoid merge commits -- do a rebase instead

**Comment by utam0k** [comment]
> @utam0k please avoid merge commits -- do a rebase instead

Done ✅

**Comment by AkihiroSuda** [comment]
@kolyshkin PTAL

**Comment by lifubang** [comment]
@utam0k Please help to do a rebase to main and solve the conflicts. Thanks.

File: docs/spec-conformance.md [hunk 21]
```diff
@@ -9,6 +9,7 @@ Spec version | Feature                                  | PR
 -------------|------------------------------------------|----------------------------------------------------------
 v1.1.0       | `SECCOMP_FILTER_FLAG_WAIT_KILLABLE_RECV` | [#3862](https://github.com/opencontainers/runc/pull/3862)
 v1.1.0       | `.process.ioPriority`                    | [#3783](https://github.com/opencontainers/runc/pull/3783)
+v1.1.0       | rsvd hugetlb cgroup                      | TODO ([#3859](https://github.com/opencontainers/runc/issues/3859))
```
Comment (AkihiroSuda) [inline docs/spec-conformance.md:12]: This should be a separate PR?

Comment (AkihiroSuda) [reply docs/spec-conformance.md:12]: Seems to have been implemented in:
- https://github.com/opencontainers/runc/pull/4073

File: libcontainer/configs/validate/validator_test.go [hunk 22]
```diff
@@ -842,3 +842,32 @@ func TestValidateScheduler(t *testing.T) {
...
+		{isErr: false, priority: 7},
+		{isErr: true, priority: -1},
+	}
+
+	for _, tc := range testCases {
+		ioPriroty := configs.IOPriority{
+			Priority: tc.priority,
+		}
+		config := &configs.Config{
+			Rootfs:     "/var",
```
Comment (AkihiroSuda) [inline libcontainer/configs/validate/validator_test.go:861]: (This value looks unusual, but it doesn’t matter here)

**Review by AkihiroSuda — APPROVED** [review]
Thanks

File: libcontainer/configs/validate/validator_test.go (hunk 22 above)
Comment (utam0k) [reply libcontainer/configs/validate/validator_test.go:861]: It seems we need to set an existing dir. I tried to delete this line and got following error.
```
 validator_test.go|869| iopriority: 0, expected nil, got error invalid rootfs: stat : no such file or directory
```

**Review by lifubang — APPROVED** [review]
Thank!
